projects/main_ping_delays_pingsender.kp/knowledge.md

```diff
@@ -0,0 +1,387 @@
+num_no_pingsender = post_no_pingsender.count()
+total_post = post_subset.count()
+num_sent_by_both =\
+    post_pingsender_only.map(lambda p: p["id"]).intersection(post_no_pingsender.map(lambda p: p["id"])).count()
+```
+We want to understand how many pings were sent by the pingsender, correctly received from the server, and sent again next time Firefox starts.
+
+
+```python
+def pct(a, b):
+    return 100 * float(a) / b
+
+print("Duplicate pings percentage: {:.2f}%".format(pct(num_sent_by_both, total_post)))
+```
+    Duplicate pings percentage: 1.38%
+
+
+Do we get many more duplicates after landing the `shutdown pingsender`?
+
+
+```python
+count_deduped_pre = dedupe(pre_subset).count()
+count_pre = pre_subset.count()
+count_deduped_post = dedupe(post_subset).count()
+count_post = post_subset.count()
+
+print("Duplicates with shutdown pingsender:\nBefore:\t{:.2f}%\nAfter:\t{:.2f}%\n"\
+      .format(pct(count_pre - count_deduped_pre, count_pre),
+              pct(count_post - count_deduped_post, count_post)))
+```
+    Duplicates with shutdown pingsender:
+    Before:	0.50%
+    After:	1.61%
+    
+
+
+It looks like 1% of the pings sent by the `pingsender` are also being sent by Firefox next time it restarts. This is potentially due to `pingsender`:
+
+- being terminated after sending the ping but before successfully deleting the ping from the disk;
+- failing to remove the ping from the disk after sending it;
+- receiving an error code from the server even when the ping was successfully sent.
+
+## Did we regress shutdownDuration?
+The `shutdownDuration` is [defined](https://gecko.readthedocs.io/en/latest/toolkit/components/telemetry/telemetry/data/main-ping.html#shutdownduration) as the time it takes to complete the Firefox shutdown process, in milliseconds. Extract the data from the two series: before the shutdown pingsender was enabled and after. Plot the data as two distinct distributions on the same plot.
+
+
+```python
+pre_shutdown_durations = pre_subset.map(lambda p: p.get("payload/simpleMeasurements/shutdownDuration", None))\
+                                   .filter(lambda p: p is not None)\
+                                   .collect()
+post_shutdown_durations = post_subset.map(lambda p: p.get("payload/simpleMeasurements/shutdownDuration", None))\
+                                     .filter(lambda p: p is not None)\
+                                     .collect()
+```
+
+```python
+plt.title("'shutdown' pingsender effect on the shutdown duration")
+plt.xlabel("shutdownDuration (milliseconds)")
+plt.ylabel("Number of pings")
+
+# Use 50 bins for values up to the clip value, and accumulate the
+# rest in the last bucket (instead of having a super-long tail).
+clip_value = 10000 # 10s
+plt.hist(np.clip(pre_shutdown_durations, 0, clip_value), alpha=0.5, bins=50, label="No pingsender")
+plt.hist(np.clip(post_shutdown_durations, 0, clip_value), alpha=0.5, bins=50, label="With pingsender")
+plt.legend()
```

Comment:
This plot's a little difficult to see. Could you make it more of a grouped barchart a la e10s_experiment.ipynb?